### PR TITLE
CI: catch ctest failures on Windows

### PR DESCRIPTION
**Pull request by ygj6**
- update `git describe ` command

 `FindGit.cmake` is copied from the Modules of cmake. It's a very old version. I am going to delete it because there is a new version in the installation directory of cmake.

The commad `${GIT_EXECUTABLE} describe --abbrev=0` in `cmake/VersionViaGit.cmake` always cause an error:
```
-- Found Git: /usr/bin/git  
fatal: No names found, cannot describe anything.
```
The reason is that the git repo in CI is not intact.

- retry cmake configuration on failure

Invalid build cache will cause cmake configuration to fail, so we need to remove the cache and retry.
On linux and macos, I used the following code:
```
cmake .. || (rm -rf * && cmake ..)
```
On windows and mingw, I use `try-catch` and `LastExitCode` to check the error and try again.

- catch ctest exception on Windows

Sometimes Powershell throws an exception, and we cannot get `$LastExitCode`.
Sometimes `$LastExitCode` is not equal to zero and Powershell will not throw an exception.
So we have to use both `$LastExitCode` and `try-catch`

- remaining issue

There is a high probability of timeout of `regress__WEPOLL`.
https://github.com/libevent/libevent/runs/715897817?check_suite_focus=true

Fixes: #1022

### CODE REVIEW

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [macos-latest]
+        os: [macos-10.15]
```
Comment (azat) [inline .github/workflows/macos.yml:28]: is this the intention? can you move this bit into separate commit?

Comment (ygj6) [reply .github/workflows/macos.yml:28]: It is to solve this problem:  https://github.com/libevent/libevent/runs/705558000?check_suite_focus=true

Comment (ygj6) [reply .github/workflows/macos.yml]: The runner has been updated from 2.169.1 to 2.263.0 by the github actions team. This will invalidate the build cache. So we need to retry as you mentioned.
